src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright 2023 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/gcs/gcs_server/gcs_autoscaler_state_manager.h"
+
+namespace ray {
+namespace gcs {
+
+GcsAutoscalerStateManager::GcsAutoscalerStateManager(
+    const ClusterResourceManager &cluster_resource_manager,
+    const GcsResourceManager &gcs_resource_manager,
+    const GcsNodeManager &gcs_node_manager)
+    : cluster_resource_manager_(cluster_resource_manager),
+      gcs_node_manager_(gcs_node_manager),
+      gcs_resource_manager_(gcs_resource_manager),
+      last_cluster_resource_state_version_(0),
+      last_seen_autoscaler_state_version_(0) {}
+
+void GcsAutoscalerStateManager::HandleGetClusterResourceState(
+    rpc::autoscaler::GetClusterResourceStateRequest request,
+    rpc::autoscaler::GetClusterResourceStateReply *reply,
+    rpc::SendReplyCallback send_reply_callback) {
+  RAY_CHECK(request.last_seen_cluster_resource_state_version() <=
+            last_cluster_resource_state_version_);
+  reply->set_last_seen_autoscaler_state_version(last_seen_autoscaler_state_version_);
+  reply->set_cluster_resource_state_version(
```

Comment:
hmm in theory this version should be bumped when state changes, not when the Get (read) request being called.

is it hard to do so?

Reply:
1. Yeah, I think it's not that straightforward: when states are "updated", I think there are a few places where we check for a flag indicating "state has changed", but in other places, we don't: we just override any cached states. (e.g. when placement group load is updated [here](https://github.com/ray-project/ray/blob/1240a81a26e379414a2e2d7119bddc61b4f007ed/src/ray/gcs/gcs_server/gcs_placement_group_manager.cc#L861)). But if we really need it, the worst we could do is to cache the last version of all states we expose, and compare the new state :) 

2. I think I am probably lacking the context of what added value it brings to the Autoscaler: within Ray, it probably doesn't matter if we bump the version with each update or not. But I understand this might make things easier for Autoscaler: e.g. autoscaler could do no-op if it sees the same version.